### PR TITLE
enable array literals (not just scalars)

### PR DESCRIPTION
**Pull request by mattjj**
Previously we only included scalars as literals in jaxprs. This PR removes that constraint, and allows arrays of any size to be included in jaxpr literals.

The reason for this change is to fix an old issue where array constants in pmapped computations were broadcasted to have a leading axis equal to the mapped axis size, e.g. from @hawkinsp:

```python
import jax
import jax.random
import numpy as np

x = jax.random.uniform(jax.random.PRNGKey(0), shape=(1000000, 10))
f = jax.pmap(lambda i: x[i])

print(jax.xla_computation(f)(np.arange(12)).GetHloText())
```

```
$ XLA_FLAGS="--xla_force_host_platform_device_count=12" python

ENTRY xla_computation_pmap__lambda___axis_name__axis_0x7fd36de1a2f0__.56 {
  ...
  constant.1 = f32[12,1000000,10]{2,1,0} constant({...})
  ...
```

The issue was that [we got consts for a staged-out pmap as outputs from a pmap bind](https://github.com/google/jax/blob/28e802c6f15d7ed87fee9bee6270137423b9967c/jax/interpreters/partial_eval.py#L151-L153), and outputs from a pmap always have a leading axis size equal to the mapped axis size (i.e. constants are broadcast). This PR changes it so that array constants aren't produced as outputs but instead immediately hoisted into the jaxpr being formed as literals. (An alternative solution would be to allow non-mapped outputs from a pmapped function, which we might want anyway, but this seems simpler for handling array constants.)

Since all JAX types are currently literalable types, I can't think of a way for the consts field to be populated now. I thought closing over outer tracers would do it, like this:

```python
@jit 
def f(x):
  @jit 
  def g():
    return x
  return g()
```

but that instead populates `env`, which means it now gets closure-converted into the formal parameters.

This choice can affect serialization, since jaxprs themselves may contain large array constants (without further processing).

TODO
- [ ] attempt to de-duplicate large literals with subjaxprs

### CODE REVIEW

File: tests/api_test.py [hunk 1]
```diff
@@ -1736,14 +1736,21 @@ def f(x):
...
+                                 in (c,) } ] b a c a d
+      in (e,) }
         """)
 
+  def test_pmap_constants_not_replicated(self):
+    x = jax.random.uniform(jax.random.PRNGKey(0), shape=(1000, 10))
+    f = jax.pmap(lambda i: x[i])
+    s = jax.xla_computation(f)(np.arange(10)).GetHloText()
+    self.assertIn('1000,10', s)
+    self.assertNotIn('1,1000,10', s)
```
Comment (gnecula) [inline tests/api_test.py]: Wouldn't this be replicated 10,1000,10, with the first 10 being the leading axis? 

This test is hard to understand. I think it would be easier if we assert on more context around '1000,10'?

**Review by gnecula — COMMENTED** [review]
The motivation and the solution seem to not match. Why were the constants hoisted out of the computation in the first place? Is that reason still applicable? 

Is there a way to deal with pmap broadcasting issue directly? If it is for the inputs, we can use the `mapped_invars` mechanism to not map the constants. Would be easy to do.

File: tests/api_test.py (hunk 1 above)
Comment (mattjj) [reply tests/api_test.py]: This is a typo on my part, and I'm confused why it isn't just crashing! I meant to give it `np.arange(1)`, so this test would run without multiple devices available.

Comment (mattjj) [reply tests/api_test.py]: I'll revise the checks to have more context.

**Comment by mattjj** [comment]
Thanks for the feedback! I should've noted that this PR is a bit of a draft at the moment, and so I'm looking for early feedback.

> Why were the constants hoisted out of the computation in the first place? Is that reason still applicable?

There wasn't a good reason; it was just an artifact of the first way we wrote things.

One benefit we realized after the fact is that it automatically had the effect of avoiding staging into XLA a large constant into both a subcomputation (like in a while_loop) and the computation that contains it. I may try to solve that more directly on this PR.

> The motivation and the solution seem to not match.

Can you say more? They seem to match in my head, but maybe I didn't explain it well!

File: tests/api_test.py (hunk 1 above)
Comment (mattjj) [reply tests/api_test.py]: I added more context to the test. As to why the pmap isn't failing when mapping over too large an axis, I believe it's because we've set up `jax.xla_computation` to work even when building computations for more devices than are available.

**Comment by gnecula** [comment]
> > The motivation and the solution seem to not match.
> 
> Can you say more? They seem to match in my head, but maybe I didn't explain it well!

What I meant was that the problem to solve is to better control how pmap broadcasts inputs and outputs. This suggests that the solution should be some control mechanism for pmap broadcasting. Incidentally, it seems that we now have the mechanism through the `mapped_invars`. 

Instead this solution is changing how we trace Python into Jaxpr and changes even how Jaxpr looks. This could have broader consequences, and also introduces a hidden dependency of pmap on how we trace constants. This reminds me very much of the freevars and how pmap was implicitly taking advantage of them, but it was not possible to see this unless you try to remove freevars.

**Comment by mattjj** [comment]
I agree that if we had both mapped_invars _and_ mapped_outvars that would give us another way of solving this. That’s what I pointed out when I wrote “An alternative solution would be...” Indeed that’s the first solution I thought of!

But that solution is heavyweight for this particular issue. Unless we have a good reason to plumb constants as outputs of a pmap, why not revise that instead? And indeed this is a revision to jaxprs, like we’ve been doing a lot of recently :) (I might try removing constants entirely now...)

This isn’t introducing a dependency of pmap on how we handle constants; you’ll notice I didn’t change the pmap code, but instead I just changed which things are hoisted as constants versus literals and everything kept working. We could flip a coin to decide whether to treat each value as a hoisted constant or a literal and it would still work. But it’s preferable to treat things as literals because it’s more efficient, hence the PR. 

Can you think of any reasons to prefer the constant-hoisting behavior?

**Comment by mattjj** [comment]
(By the way, handling a pmap with mapped_outvars is tricky because, for example, when doing jvp or partial eval we don’t know until we run the function how many outputs there will be. It’s doable but much heavier than anything in this PR!)

**Comment by gnecula** [comment]
To be clear, if you say there were no important reasons to hoist constants, I like the change. If we can remove hoisted consts, I think that Jaxpr will be even simpler :-) 

I see that you did not change the code for pmap, yet you fixed how it behaves. This suggests that the dependency is implicit. I would have thought that with or without hoisting the semantics should be the same.

**Review by gnecula — APPROVED** [review]

**Comment by gnecula** [comment]
If you do agree to try to remove the constvars, then I'd be happy to do that. I assume that some work would be needed for the linearity-bitmaps for scan and while, and also maybe something for compilation (perhaps hoisting them out only for compilation)?

File: jax/interpreters/partial_eval.py [hunk 2]
```diff
@@ -74,7 +74,7 @@ def instantiate_const(self, tracer):
     if isinstance(pv, AbstractValue):
       return tracer
     elif pv is None:
-      if type(const) in core.literalable_types and onp.shape(const) == ():
```
Comment (gnecula) [inline jax/interpreters/partial_eval.py]: I am curious, what are non core.literalable_types. In other words, what things will still be lifted as consts?

**Comment by jakevdp** [comment]
Closing, as non-scalar literals are now supported!
